metadatastore/commands.py

```diff
@@ -541,13 +550,137 @@ def _get_mongo_document(document, document_cls):
+        collection
+    start_time : time-like, optional
+        time-like representation of the earliest time that a RunStart
+        was created. Valid options are:
+           - timestamps --> time.time()
+           - '2015'
+           - '2015-01'
+           - '2015-01-30'
+           - '2015-03-30 03:00:00'
+           - datetime.now()
+    stop_time : time-like, optional
+        timestamp of the latest time that a RunStart was created. See
+        docs for `start_time` for examples.
+    uid : str, optional
+        Globally unique id string provided to metadatastore
+    """
+    _normalize_object_id(kwargs, '_id')
+    _format_time(kwargs)
+    correction_dicts = _find_documents(Correction, **kwargs)
+    for correction_dict in correction_dicts:
+        if dereference_uids:
+            _dereference_reference_fields(correction_dict, newest=newest)
+        yield Document(correction_dict, correction_dict[
+            'original_document_type'] + ' -- Correction')
+
+
+def _dereference_reference_fields(mongo_dict, newest=True):
+    """In-place dereferencing
+    """
+    # reference_fields = {'run_start_id': find_run_starts,
+    #                     'run_start': find_run_starts,
+    #                     'descriptor_id': find_event_descriptors,
+    #                     'descriptor': find_event_descriptors,
+    #                     'beamline_config_id': find_beamline_configs,
+    #                     'beamline_config': find_beamline_configs}
+    reference_fields = {'run_start_id': RunStart,
+                        'run_start': RunStart,
+                        'descriptor_id': EventDescriptor,
+                        'descriptor': EventDescriptor,
+                        'beamline_config_id': BeamlineConfig,
+                        'beamline_config': BeamlineConfig}
+
+    for ref, klass in six.iteritems(reference_fields):
+        field = mongo_dict.get(ref, None)
+        if field:
+            # then field is either an ObjectId or a uid, or the object model
+            # is borked...
+            if isinstance(field, ObjectId):
+                # find the original document
+                kwargs = {'_id': field}
+            elif isinstance(field, six.text_type):
+                # it better be a uid
+                kwargs = {'uid': field}
+            else:
+                break
+            document_generator = _find_documents(klass, **kwargs)
+            document = next(document_generator)
+            document = Document(document, klass.__name__)
+            if newest and 'beamline_config' not in ref:
+                corrected_document_generator = find_corrections(
+                    uid=document['uid'])
+                try:
+                    document = next(corrected_document_generator)
+                except StopIteration:
+                    pass
+            _dereference_reference_fields(document, newest=newest)
+            if ref.endswith('_id'):
```

Comment:
Can you add a comment about why this is being done so we remember next time we read this.